README
NAME
    Exception::System - The exception class for system or library calls

SYNOPSIS
      # The simplest usage
      use Exception::Base 'Exception::System';
      open my $file, "/notfound"
        or Exception::System->throw(message=>"Can not open file");

      # The Exception::System class can be a base class for others
      #
      # Loaded automatically if used as Exception::Base's argument
      use Exception::Base,
        'Exception::System',
        'Exception::File' => {
            isa => 'Exception::System',
            has => 'filename',
            stringify_attributes => [ 'message', 'errstr', 'filename' ],
        };

      eval {
        my $filename = "/notfound";
        open my $fh, $filename
            or Exception::File->throw(
                   message=>"Can not open file",
                   filename=>$filename,
               );
      };
      if ($@) {
        my $e = Exception::Base->catch;
        if ($e->isa('Exception::File')) { warn "File error:".$e->errstr; }
        if ($e->matches({errname=>'ENOENT'})) { warn "Caught not found error"; }
      }

DESCRIPTION
    This class extends standard Exception::Base with handling system or
    library errors. The additional attributes of the exception object are
    filled on throw and contain the error message and error codes.

AUTHOR
    Piotr Roszatycki <dexter@debian.org>

LICENSE
    Copyright (C) 2007, 2008 by Piotr Roszatycki <dexter@debian.org>.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://www.perl.com/perl/misc/Artistic.html



Changes
------------------------------------------------------------------------
0.1001 | Piotr Roszatycki <dexter@debian.org> | 2008-10-13

Changes:

* Requires Exception::Base >= 0.19.

* Build requires Test::Unit::Lite >= 0.10.

* Minor POD changes.

------------------------------------------------------------------------
0.10 | Piotr Roszatycki <dexter@debian.org> | 2008-10-07

Changes:

* "strno" attribute represents numeric value of object.

  eval { open my $f, "/file" or Exception::System->throw };
  if ($@ and $@ == &Errno::ENOENT) { warn "file not found" }

* Uses "numeric_attribute" and "stringify_attributes" attributes.

* Requires Exception::Base >= 0.18.

------------------------------------------------------------------------
0.0902 | Piotr Roszatycki <dexter@debian.org> | 2008-05-29

Changes:

* Fixed POD coverage.

------------------------------------------------------------------------
0.0901 | Piotr Roszatycki <dexter@debian.org> | 2008-05-08

Changes:

* Fixed MANIFEST.

------------------------------------------------------------------------
0.09 | Piotr Roszatycki <dexter@debian.org> | 2008-05-08

Changes:
------------------------------------------------------------------------
0.09 | Piotr Roszatycki <dexter@debian.org> | 2008-05-08

Changes:

* Requires Exception::Base >= 0.16.
* Build requires Test::Unit::Lite >= 0.07.
* The FIELD constant renamed to ATTRS as far as every OO language calls it
  attributes.
* See Exception::Base for incompatibilities list.

------------------------------------------------------------------------
0.0802 | Piotr Roszatycki <dexter@debian.org> | 2008-04-19

Changes:

* Requires Exception::Base >= 0.15.
* Drop indirect notation is POD and code.

------------------------------------------------------------------------
0.0801 | Piotr Roszatycki <dexter@debian.org> | 2008-04-13

Changes:

* Prerequisites Exception::Base >= 0.14.
* Updated bundled Test::Unit::Lite to 0.0903.

------------------------------------------------------------------------
0.08 | Piotr Roszatycki <dexter@debian.org> | 2007-11-05

Changes:

* Get shorter errname string for the same errno number.  It should give the
  same errname strings on Unix and Windows.
* Updated bundled Test::Unit::Lite to 0.0701.

------------------------------------------------------------------------
0.0702 | Piotr Roszatycki <dexter@debian.org> | 2007-10-29

Changes:

* Updated bundled Test::Unit::Lite to 0.07.
* More strict test units.

------------------------------------------------------------------------
0.0701 | Piotr Roszatycki <dexter@debian.org> | 2007-10-09

Changes:

* Do not test POD errors unless $ENV{PERL_TEST_POD} is true.

------------------------------------------------------------------------
0.07 | Piotr Roszatycki <dexter@debian.org> | 2007-10-08

New:

* Defined accessors.

Changes:

* Switched to Module::Build.
* Updated bundled Test::Unit::Lite to 0.05.
* Prerequisites Exception::Base >= 0.09
* The err* fields are reset after calling _collect_system_data().
* Returns errstr only if message is empty.

------------------------------------------------------------------------
0.0601 | Piotr Roszatycki <dexter@debian.org> | 2007-05-18

* Prerequisites Exception::Base >= 0.06

------------------------------------------------------------------------
0.06 | Piotr Roszatycki <dexter@debian.org> | 2007-05-18

* Bundle Test::Unit::Lite 0.02 to fix compilation on Perl 5.006 and 5.009.
* Change default message to 'Unknown system exception'.

------------------------------------------------------------------------
0.05 | Piotr Roszatycki <dexter@debian.org> | 2007-05-11

* Added bundled Test/Unit/Lite.pm to MANIFEST.

------------------------------------------------------------------------
0.04 | Piotr Roszatycki <dexter@debian.org> | 2007-05-09

* use 5.006
* Test with bundled Unit::Test::Lite, so there is no dependency on
  Unit::Test.
* META.yml is generated with Makefile.PL.

------------------------------------------------------------------------
0.03 | Piotr Roszatycki <dexter@debian.org> | 2007-04-24

* require 5.006
* META.yml created by hand.

------------------------------------------------------------------------
0.02 | Piotr Roszatycki <dexter@debian.org> | 2007-04-22

* First CPAN release.

------------------------------------------------------------------------
0.01 | Piotr Roszatycki <dexter@debian.org> | 2007-04-20

* Development release.

------------------------------------------------------------------------


t/profile.pl
#!/usr/bin/perl -d:DProf

use lib 'lib', '../lib';	
use Exception::Base ':all',
    'Exception::System';

my $n = 0;

foreach (1..10000) {
    try eval { Exception::System->throw(message=>'Message') };
    if (catch my $e) {
        if ($e->isa('Exception::System') and $e->with('Message')) { $n++; }
    }
}

print "tmon.out data collected. Call dprofpp\n";


t/benchmark.pl
#!/usr/bin/perl -al

BEGIN {
    {
	package My::Eval;
	our $n = 0;
	sub test {
	    eval { 1; };
	    $n++;
	}
    }

    {
	package My::DieScalar;
	our $n = 0;
	sub test {
	    eval { die "Message\n"; };
	    if ($@ eq "Message\n") { $n++; }
	}
    }

    {
	package My::DieObject;
	our $n = 0;
	sub test {
	    eval { My::DieObject->throw };
	    if ($@ and $@->isa('My::DieObject')) { $n++; }
	}
	sub throw {
	    my %args = @_;
	    die bless {%args}, shift;
	}
    }
    
    {
	package My::ExceptionBase;
	use lib '../lib';	
	use Exception::Base ':all';
	our $n = 0;
	sub test {
	    try eval { throw 'Exception::Base' => message=>'Message'; };
	    if (catch my $e) {
	        if ($e->isa('Exception::Base') and $e->with('Message')) { $n++; }
	    }
	}
    }

    {
	package My::ExceptionBase1;
	use lib 'lib';	
	use Exception::Base;
	our $n = 0;
	sub test {
	    Exception::Base->try(eval { Exception::Base->throw(message=>'Message', verbosity=>1); });
	    if (Exception::Base->catch(my $e)) {
	        if ($e->isa('Exception::Base') and $e->with('Message')) { $n++; }
	    }
	}
    }

    {
	package My::ExceptionSystem;
	use lib 'lib', '../lib';
	use Exception::Base ':all', 'Exception::System';
	our $n = 0;
	sub test {
	    try eval { throw 'Exception::System' => message=>'Message'; };
	    if (catch my $e) {
	        if ($e->isa('Exception::System') and $e->with('Message')) { $n++; }
	    }
	}
    }

    {
	package My::ExceptionSystem1;
	use lib 'lib', '../lib';
	use Exception::System;
	our $n = 0;
	sub test {
	    Exception::System->try(eval { Exception::System->throw(message=>'Message', verbosity=>1); });
	    if (Exception::System->catch(my $e)) {
	        if ($e->isa('Exception::System') and $e->with('Message')) { $n++; }
	    }
	}
    }

    eval q{
	package My::ErrorSystemException;
	use Error qw(:try);
	use Error::SystemException;
	our $n = 0;
	sub test {
	    try {
	        throw Error::SystemException('Message');
	    }
	    catch Error with {
	        my $e = shift;
	        if ($e->text eq 'Message') { $n++; }
	    };
	}
    };
}

package main;

use Benchmark ':all';

my %tests = (
    '1_DieScalar'               => sub { My::DieScalar::test; },
    '2_DieObject'               => sub { My::DieObject::test; },
    '3_ExceptionBase'           => sub { My::ExceptionBase::test; },
    '4_ExceptionBase1'          => sub { My::ExceptionBase1::test; },
    '5_ExceptionSystem'         => sub { My::ExceptionSystem::test; },
    '6_ExceptionSystem1'        => sub { My::ExceptionSystem1::test; },
);
$tests{'7_ErrorSystemException'} = sub { My::ErrorSystemException::test; } if eval { Error->VERSION };

my $result = timethese(-1, { %tests });
cmpthese($result);


t/profile_try.pl
#!/usr/bin/perl -d:DProf

use lib 'lib', '../lib';	
use Exception::Base 'Exception::System';

foreach (1..10000) {
    try eval { Exception::System->throw(message=>'Message') };
    if (catch my $e) {
        if ($e->isa('Exception::Base') and $e->with('Message')) { 1; }
    }
}

print "tmon.out data collected. Call dprofpp\n";


eg/file-not-found_exception.pl
#!/usr/bin/perl -I../lib

# It is almost the same example as for Exception::Base. In this one the
# Exception::IO is a Exception::System.

# Use module and create needed exceptions
use Exception::Base
    'Exception::IO' => { isa => 'Exception::System' },
    'Exception::FileNotFound' => {
	isa => 'Exception::IO',
	has => 'filename',
	stringify_attributes => [ 'message', 'errstr', 'filename' ],
    };

sub func1 {
    eval {
        my $filename = '/notfound';
        open my $fh, $filename
            or throw Exception::FileNotFound
                     message=>'Can not open file', filename=>$filename;
    };

    if ($@) {
	my $e = Exception::IO->catch;
        # $e is an exception object for sure, no need to check if is blessed
        warn "*** Exception caught";
        if ($e->isa('Exception::FileNotFound')) {
            warn "*** Exception caught for file " . $e->filename
               . " with error " . $e->errname;
        }
        # rethrow the exception
        warn "*** Rethrow exception";
        $e->throw;
    }
}

sub func2 {
    func1(1);
}

sub func3 {
    func2(2,2);
}

func3(3,3,3);


lib/Exception/System.pm
#!/usr/bin/perl -c

package Exception::System;
use 5.006;
our $VERSION = 0.10_01;

=head1 NAME

Exception::System - The exception class for system or library calls

=head1 SYNOPSIS

  # The simplest usage
  use Exception::Base 'Exception::System';
  open my $file, "/notfound"
    or Exception::System->throw(message=>"Can not open file");

  # The Exception::System class can be a base class for others
  #
  # Loaded automatically if used as Exception::Base's argument
  use Exception::Base,
    'Exception::System',
    'Exception::File' => {
        isa => 'Exception::System',
        has => 'filename',
        stringify_attributes => [ 'message', 'errstr', 'filename' ],
    };

  eval {
    my $filename = "/notfound";
    open my $fh, $filename
        or Exception::File->throw(
               message=>"Can not open file",
               filename=>$filename,
           );
  };
  if ($@) {
    my $e = Exception::Base->catch;
    if ($e->isa('Exception::File')) { warn "File error:".$e->errstr; }
    if ($e->matches({errname=>'ENOENT'})) { warn "Caught not found error"; }
  }

=head1 DESCRIPTION

This class extends standard L<Exception::Base> with handling system or library
errors.  The additional attributes of the exception object are filled on throw
and contain the error message and error codes.

=for readme stop

=cut


use strict;
use warnings;


# Extend Exception::Base class
use Exception::Base 0.19;
use base 'Exception::Base';


# Use ERRNO hash
use Errno ();


# List of class attributes (name => {is=>ro|rw, default=>value})
use constant ATTRS => {
    %{ Exception::Base->ATTRS },     # SUPER::ATTRS
    stringify_attributes => { default => [ 'message', 'errstr' ] },
    numeric_attribute    => { default => 'errno' },
    message  => { is => 'rw', default => 'Unknown system exception' },
    errstr   => { is => 'ro' },
    errstros => { is => 'ro' },
    errno    => { is => 'ro' },
    errname  => { is => 'ro' },
};


# Map for errno -> errname (choose the shortest errname string for the same errno number)
my %Errname = map { Errno->$_ => $_ }
              sort { length $b <=> length $a }
              sort
              keys (%!);


# Collect system data
sub _collect_system_data {
    my $self = shift;

    $self->{errstr}   = "$!";   # string context
    $self->{errstros} = $^E;
    $self->{errno}    = 0+$!;   # numeric context
    $self->{errname}  = $Errname{ $self->{errno} } || '';

    return $self->SUPER::_collect_system_data(@_);
}


# Module initialization
sub __init {
    __PACKAGE__->_make_accessors;
}


__init;


1;


__END__

=begin umlwiki

= Class Diagram =

[                       <<exception>>
                      Exception::System
 -------------------------------------------------------------
 +message : Str = "Unknown system exception"             {new}
 +errstr : Str
 +errstros : Str
 +errno : Int
 +errname : Str
 #numeric_attribute : Str = "strno"
 #stringify_attributes : ArrayRef[Str] = ["message", "errstr"]
 -------------------------------------------------------------
 #_collect_system_data()
 <<constant>> +ATTRS() : HashRef                              ]

[Exception::System] ---|> [Exception::Base]

=end umlwiki

=head1 BASE CLASSES

=over

=item *

L<Exception::Base>

=back

=head1 CONSTANTS

=over

=item ATTRS

Declaration of class attributes as reference to hash.

See L<Exception::Base> for details.

=back

=head1 ATTRIBUTES

Class attributes are implemented as values of blessed hash.  The attributes of
base class are inherited.  See L<Exception::Base> to see theirs description.

=over

=item message (rw, default: 'Unknown system exception')

Contains the message of the exception.  This class overrides the default value
from L<Exception::Base> class.

=item errstr (ro)

Contains the system error string fetched at exception throw.  It is the part
of the string representing the exception object.  It is the same as B<$!>
variable in string context.

  eval { Exception::System->throw( message=>"Message" ) };
  my $e = Exception::Base->catch
    and print $e->errstr;

=item errstros (ro)

Contains the extended system error string fetched at exception throw.  It is
the same as B<$^E> variable.

  eval { Exception::System->throw( message=>"Message" ); };
  if ($@) {
    my $e = Exception::Base->catch;
    if ($e->errstros ne $e->errstr) {
      print $e->errstros;
    }
  }

=item errno (ro)

Contains the system error number fetched at exception throw.  It is the same
as B<$!> variable in numeric context.  This attribute represents numeric value
of the exception object in numeric context.

  use Errno ();
  eval { Exception::System->throw( message=>"Message" ); };
  if ($@) {
    my $e = Exception::Base->catch;
    if ($e->errno == &Errno::ENOENT) {  # explicity
      warn "Not found";
    }
    elsif ($e == &Errno::EPERM) {       # numeric context
      warn "Bad permissions";
    }
  }

=item errname (ro)

Contains the system error constant from the system F<error.h> include file.

  eval { Exception::System->throw( message=>"Message" ); };
  my $e = Exception::Base->catch
    and $e->errname eq 'ENOENT'
    and $e->throw;

=item stringify_attributes (default: ['message', 'errstr'])

Meta-attribute contains the format of string representation of exception
object.  This class overrides the default value from L<Exception::Base>
class.

=item numeric_attribute (default: 'errno')

Meta-attribute contains the name of the attribute which contains numeric
value of exception object.  This class overrides the default value from
L<Exception::Base> class.

=back

=head1 PRIVATE METHODS

=over

=item _collect_system_data

Collect system data and fill the attributes of exception object.  This
method is called automatically if exception if throwed.  This class
overrides the method from L<Exception::Base> class.

=back

=head1 SEE ALSO

L<Exception::Base>.

=head1 BUGS

If you find the bug, please report it.

=for readme continue

=head1 AUTHOR

Piotr Roszatycki E<lt>dexter@debian.orgE<gt>

=head1 LICENSE

Copyright (C) 2007, 2008 by Piotr Roszatycki E<lt>dexter@debian.orgE<gt>.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>
